ark/jsonschema/array.ts

```diff
@@ -0,0 +1,159 @@
+import type { Type, applyConstraintSchema } from "arktype"
+
+import { innerParseJsonSchema, type inferJsonSchema } from "./json.ts"
+import { JsonSchema } from "./scope.ts"
+
+const deepNormalize = (data: unknown): unknown =>
+	typeof data === "object" ?
+		data === null ? null
+		: Array.isArray(data) ? data.map(item => deepNormalize(item))
+		: Object.fromEntries(
+				Object.entries(data)
+					.map(([k, v]) => [k, deepNormalize(v)] as const)
+					.sort((l, r) => (l[0] > r[0] ? 1 : -1))
+			)
+	:	data
+
+const arrayItemsAreUnique = (
+	array: readonly unknown[],
+	ctx: TraversalContext
+) => {
+	const seen: Record<string, true> = {}
+	const duplicates: unknown[] = []
+	for (const item of array) {
+		const stringified = JSON.stringify(deepNormalize(item))
+		if (stringified in seen) duplicates.push(item)
+		else seen[stringified] = true
+	}
+	return duplicates.length === 0 ?
+			true
+		:	ctx.reject({
+				expected: "unique array items",
+				actual: `duplicated at elements ${printable(duplicates)}`
+			})
+}
+
+const arrayContainsItemMatchingSchema = (
+	array: readonly unknown[],
+	schema: Type<unknown>,
+	ctx: TraversalContext
+) =>
+	array.some(item => schema.allows(item)) === true ?
+		true
+	:	ctx.mustBe(
+			"an array containing at least one item matching 'contains' schema"
+		)
+
+export const validateJsonSchemaArray = JsonSchema.ArraySchema.pipe(
+	jsonSchema => {
+		const arktypeArraySchema: Intersection.Schema<Array<unknown>> = {
+			proto: "Array"
+		}
+
+		if ("items" in jsonSchema) {
+			if (Array.isArray(jsonSchema.items)) {
+				arktypeArraySchema.sequence = {
+					prefix: jsonSchema.items.map(
+						item => innerParseJsonSchema.assert(item).internal
+					)
+				}
+
+				if ("additionalItems" in jsonSchema) {
+					if (jsonSchema.additionalItems === false)
+						arktypeArraySchema.exactLength = jsonSchema.items.length
+					else {
+						arktypeArraySchema.sequence = {
+							...arktypeArraySchema.sequence,
+							variadic: innerParseJsonSchema.assert(jsonSchema.additionalItems)
+								.internal
+						}
+					}
+				}
+			} else {
+				arktypeArraySchema.sequence = {
+					variadic: innerParseJsonSchema.assert(jsonSchema.items).internal
+				}
+			}
+		}
+
+		if ("maxItems" in jsonSchema)
+			arktypeArraySchema.maxLength = jsonSchema.maxItems
+		if ("minItems" in jsonSchema)
+			arktypeArraySchema.minLength = jsonSchema.minItems
+
+		const predicates: Predicate.Schema[] = []
+		if ("uniqueItems" in jsonSchema && jsonSchema.uniqueItems === true)
+			predicates.push((arr: unknown[], ctx) => arrayItemsAreUnique(arr, ctx))
+
+		if ("contains" in jsonSchema) {
+			const parsedContainsJsonSchema = innerParseJsonSchema.assert(
+				jsonSchema.contains
+			)
+			predicates.push((arr: unknown[], ctx) =>
+				arrayContainsItemMatchingSchema(arr, parsedContainsJsonSchema, ctx)
+			)
+		}
+
+		arktypeArraySchema.predicate = predicates
+
+		return rootSchema(arktypeArraySchema) as unknown as Type<JsonSchema.Json[]>
```

Comment:
If you ever are casting a return multiple times like this, best to just cast it to `never` and annotate the return type

Reply:
Fixed this and other instances of the same thing in 01e296c2